svc/svc.go

```diff
@@ -0,0 +1,246 @@
+//
+// Recommend Usage:
+//
+//	func main() {
+//		env := svc.InitAll()
+//		defer env.Close()
+//
+//		r := httpx.NewRouter()
+//		// ... add routes
+//
+//		h := svc.NewStandardHandler(svc.HandlerOpts{
+//			Router:   r,
+//			Reporter: env.Reporter,
+//	})
+//
+// 	svc.RunServer(h, "80", 5*time.Second)
+// }
+package svc
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+	"os/signal"
+	"strings"
+	"syscall"
+	"time"
+
+	ddtrace "github.com/DataDog/dd-trace-go/opentracing"
+	opentracing "github.com/opentracing/opentracing-go"
+	"github.com/pkg/errors"
+	"github.com/remind101/pkg/httpx"
+	"github.com/remind101/pkg/httpx/middleware"
+	"github.com/remind101/pkg/logger"
+	"github.com/remind101/pkg/metrics"
+	"github.com/remind101/pkg/reporter"
+	"github.com/remind101/pkg/reporter/rollbar"
+)
+
+type HandlerOpts struct {
+	Router            *httpx.Router
+	Reporter          reporter.Reporter
+	ForwardingHeaders []string
+	BasicAuth         string
+}
+
+// NewStandardHandler returns an http.Handler with a standard middleware stack.
+// The last middleware added is the first middleware to handle the request.
+// Order is pretty important as some middleware depends on others having run
+// already.
+func NewStandardHandler(opts HandlerOpts) http.Handler {
+	var h httpx.Handler
+
+	// Recover from panics. A panic is converted to an error. This should be first,
+	// even though it means panics in middleware will not be recovered, because
+	// later middleware expects endpoint panics to be returned as an error.
+	h = middleware.BasicRecover(opts.Router)
+
+	// Add request tracing. Must go before the HandleError middleware in order
+	// to capture any errors from the endpoint handler.
+	h = middleware.OpentracingTracing(h, opts.Router)
+
+	// Handler errors returned by endpoint handler or recovery middleware.
+	// Errors will no longer be returned after this middeware.
+	h = middleware.HandleError(h, middleware.ReportingErrorHandler)
+
+	// Insert logger into context and log requests at INFO level.
+	h = middleware.LogTo(h, middleware.LoggerWithRequestID)
+
+	// Add reporter to context and request to reporter context.
+	h = middleware.WithReporter(h, opts.Reporter)
+
+	// Add the request id to the context.
+	h = middleware.ExtractRequestID(h)
+
+	// Add basic auth
+	if opts.BasicAuth != "" {
+		user := strings.Split(opts.BasicAuth, ":")[0]
+		pass := strings.Split(opts.BasicAuth, ":")[1]
+		h = middleware.BasicAuth(h, user, pass, "")
+	}
+
+	// Adds forwarding headers from request to the context. This allows http clients
+	// to get those headers from the context and add them to upstream requests.
+	if len(opts.ForwardingHeaders) > 0 {
+		for _, header := range opts.ForwardingHeaders {
+			h = middleware.ExtractHeader(h, header)
+		}
+	}
+
+	// Wrap the route in middleware to add a context.Context. This middleware must be
+	// last as it acts as the adaptor between http.Handler and httpx.Handler.
+	return middleware.BackgroundContext(h)
```

Comment:
actually, just was playing with this in permissions and this means that you can't add any other middleware since this adapts the `httpx.Handler` into an `http.Handler`. Would like for this to just return an `httpx.Handler` and have the client of this figure out the rest

Reply:
Yeah, that's one issue with this approach, its a bit of a "take it or leave it" stack. Even if we returned an httpx.Handler though, it would be hard to customize since you wouldn't be able to insert a middleware where you need it.

We can add things that need customization to the opts, for example:
``` golang
type HandlerOpts struct {
	Router            *httpx.Router
	Reporter          reporter.Reporter
	ForwardingHeaders []string
	BasicAuth         string

        // Allow customizing error handling func
        ErrorHandlerFunc middleware.ErrorHandlerFunc
}
```

Error handling seems like it'll be a pretty common one that'll need customizing, at least the response formatting part of it.

Another idea, we could allowing middleware insertion in a good spot in the stack.
``` golang
type HandlerOpts struct {
	Router            *httpx.Router
	Reporter          reporter.Reporter
	ForwardingHeaders []string
	BasicAuth         string
	Middlewares       []Middleware
}

type Middleware func(httpx.Handler) httpx.Handler

func NewStandardHandler(opts HandlerOpts) http.Handler {
	var h httpx.Handler
	h = middleware.BasicRecover(opts.Router)

	// Add custom middleware here.
	for _, m := range opts.Middlewares {
		h = m(h)
	}

	h = middleware.OpentracingTracing(h, opts.Router)
	h = middleware.HandleError(h, middleware.ReportingErrorHandler)
	h = middleware.LogTo(h, middleware.LoggerWithRequestID)
	h = middleware.WithReporter(h, opts.Reporter)
	h = middleware.ExtractRequestID(h)
	if opts.BasicAuth != "" {
		user := strings.Split(opts.BasicAuth, ":")[0]
		pass := strings.Split(opts.BasicAuth, ":")[1]
		h = middleware.BasicAuth(h, user, pass, "")
	}
	if len(opts.ForwardingHeaders) > 0 {
		for _, header := range opts.ForwardingHeaders {
			h = middleware.ExtractHeader(h, header)
		}
	}
	return middleware.BackgroundContext(h)
}
```

Reply:
It sounds like we have a canonical mapping for `httpx.Handler -> http.Handler`. That's `middleware.BackgroundContext`.

What if we added `func (h *httpx.Handler) HttpHandler() http.Handler`?

Then in services using this, they'd do something like:

```
http.Handle("/some_path", httpx.NewStandardHandler(opts).HttpHandler())
```
unless they wanted to  attach additional middleware, then they could do:
```
h := httpx.NewStandardHandler(opts)
h = middleware.SomeNonStandardMiddleware(h)
http.Handle("/some_path", h.HttpHandler())
```

Reply:
Its true that `middleware.BackgroundContext` is really the adapter between httpx and http handlers. It could be renamed to `httpx.HTTPHandler` even. It doesn't even generate a background context anymore. It uses `request.Context()`. 

I don't know if adding `HttpHandler() http.Handler` to the httpx.Handler interface is worth it. That means every httpx middleware needs to add that function, which will be the same thing in every case. 

I'd like to see us need to add custom middleware at the top of the stack before we make a change like this. My guess is we won't need to do this, or if we find a new middleware we want to add, it'll make sense to make it an option in the HandlerOpt struct.

Reply:
We should probably grab some time monday to chat in person about this one. I'm coming around to leaving as is and passing in custom handlers in as opposed to wrapping the output of this in custom handlers, especially given the importance of ordering. But lets get the ideas out on the table and come to some consensus